### ISSUE TITLE
Messages are received out of order for partitions

### ISSUE THREAD

**Issue by anurag108** [SOHU-Co/kafka-node#788]
* Node version: v8.2.1
* Kafka-node version: 2.2.3
* Kafka version: 0.8.2.2
* Number partitions for topic: 16

NOTE: I can not upgrade my kafka cluster to 0.9+, so I wont be able to use ConsumerStream or ConsumerGroup

### Issue Description
It is mentioned that messages are emitted synchronously here https://github.com/SOHU-Co/kafka-node/issues/707#issuecomment-314479483.

So it is natural to assume that we'll get messages per partition in order. But thats not what is happening. I am able to reproduce this over and over again. I have attached the code to reproduce this behaviour and have attached one of the sample output I got as well.

Another issue is that I am explicitly committing the offsets after whatever work I need to do with the messages. And I am also pausing the consumer as soon as I receive a message so that no more fetch loops are executed. But when I resume the consumer, I do not get messages from the offsets I committed. Sample output for this issue is also attached. As you can see for partition 8 even after repeatedly committing an offset, I continue to receive messages from subsequent offsets.

### Sample Code to reproduce behavior
```js
const kafka = require('kafka-node');
const client = new kafka.Client('my_host:2181');
const offset = new kafka.Offset(client);

const topic = 'my_topic';
const groupId = 'random_group_id_001';

const consumer = new kafka.HighLevelConsumer(client, [{
  topic,
}], {
  groupId,
  autoCommit: false,
  fromOffset: false,
  fetchMaxBytes: 5 * 1024,
});

let waitTimeout;
let startedWork = false;
const offsets = {};
let messageArray = [];

function waitForSometime(time) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, time);
  });
}

function commitOffsets() {
  return new Promise((resolve, reject) => {
    const partitions = Object.keys(offsets);
    const payloads = [];
    for (let i = 0; i < partitions.length; i += 1) {
      payloads.push({
        topic,
        partition: partitions[i],
        offset: offsets[i],
      });
      console.log(`${Date.now()} committing offset ${partitions[i]} ${offsets[partitions[i]]}`);
    }
    offset.commit(groupId, payloads, (error, data) => (error ? reject(error) : resolve(data)));
  });
}

function setNewTimeout() {
  if (waitTimeout) {
    clearTimeout(waitTimeout);
  }
  waitTimeout = setTimeout(async () => {
    startedWork = true;
    await waitForSometime(2000);
    await commitOffsets();
    messageArray = [];
    startedWork = false;
    consumer.resume();
  }, 1000);
}

consumer.on('message', async (message) => {
  setNewTimeout();
  consumer.pause();
  if (!startedWork) {
    if (offsets[message.partition]) {
      if ((offsets[message.partition] + 1) !== message.offset) {
        console.log(`${Date.now()} offset mismatch for partition ${message.partition}. found: ${message.offset} expected ${offsets[message.partition]}`);
      } else {
        messageArray.push(message);
        offsets[message.partition] = message.offset + 1;
        console.log(`${Date.now()} got message ${message.partition} ${message.offset}`);
      }
    } else {
      // should work since we should get messages from each partition in order
      console.log(`${Date.now()} got message ${message.partition} ${message.offset}`);
      offsets[message.partition] = message.offset + 1;
    }
  } else {
    console.log(`${Date.now()} message received after timeout ${message.partition} ${message.offset}`);
  }
});
```

### Out of Order Sample Output
[SampleOutput.txt](https://github.com/SOHU-Co/kafka-node/files/1414153/SampleOutput.txt)

### Offset Mismatch Sample Output
[offset_mismatch_sample.txt](https://github.com/SOHU-Co/kafka-node/files/1414324/offset_mismatch_sample.txt)

**Comment by hyperlink** [comment]
`HighLevelConsumer` keeps track of it's own offsets locally and so if you are committing outside of it is not going to be aware of that.

**Comment by anurag108** [comment]
@hyperlink so say I want to get some work done with the messages I've got and I want to commit the offsets only when I have completed that work successfully. I cannot set autoCommit to true because if my worker fails I would need to reprocess those messages. How do you suggest commit the offsets only after the work is done?

Also what about messages arriving out of order even with each partition. That shouldn't be happening, right?

**Comment by hyperlink** [comment]
Yeah it's a common use case which is why ConsumerGroupStream to only commits what was consumed from the buffer. Maybe look at the code there and build a similar solution?

Regarding the out of order messages I believe that can happen when compression is enabled but I have not had a chance to look into it.

**Comment by anurag108** [comment]
Cool, thanks for this clarification
